### PR TITLE
Add cron.deny/allow and refactored for Puppet4

### PR DESCRIPTION
**Pull request by crayfishx**
This is a fairly extensive refactor to get rid of some of the unsupported stuff (like using hiera_hash(), create_resources..etc).  I also pulled in #24 into this PR as @pillarsdotnet had already added the type casting to parameters

Closes #24 

###  Refactored for Puppet 4 and added cron.allow/deny

    * Feature: add management of /etc/cron.allow and /etc/cron.deny
    * Enhancement: Replaced hiera_hash() with lookup() calls
    * Enhancement: Replaced params.pp with in-module data (Hiera 5)
    * Enhancement: Replaced create_resources with iterators
    * Enhancement: Replaced anchor pattern with contain
    * Metadata: Changed compatibility to Puppet >4.9.0 <6.0.0

### Feature: /etc/cron.allow,deny
This PR adds the ability to manage `/etc/cron.allow` and `/etc/cron.deny` using the `$users_allow` and `$users_deny` parameters.  It can be switched on and off with the `$manage_users_allow` and `$manage_users_deny` parameters.  Both of these params currently default to `false` to help with backwards compatibility - so this feature is currently opt-in

### Replaced hiera_hash()
`hiera_hash()` has reached the end of the road - this has been replaced with the `lookup()` function  introduced the Puppet 4.x series

### Replaced params.pp pattern
The params.pp pattern is no longer the necessary evil it once was.  This pattern is now considered dated and this PR replaces params.pp with the new style data-in-modules using Hiera 5

### Replaced `create_resources()`
Like params, `create_resources()` is something that just needs to go away now - this PR replaces the `create_resources` calls with iterators.

### Replaced anchors with contain
The anchor pattern was something of a hack that is now achievable in a much nicer way using `contain`

## Compatibility

Technically, due to the use of Hiera 5 these changes mean the module is now compatible out-of-the-box with Puppet 4.9 ( < 6.0.0 ).  Users of older Puppet 4 installation can still get it working though, they'd just need to add some extra Hiera data to account for data in modules not being available.  I think this could easily be documented, and there should be very few reasons why existing Puppet 4 users shouldn't be updating to at least 4.10.  This module is still 0.x, I dont think we should be afraid to add the dependency on 4.9+ - it's better than starting out on unsupported patterns.

### CODE REVIEW

**Comment by roman-mueller** [comment]
I love this but the hard dependency to 4.9+ scares me.
I know this module is only version 0.x but for some reason it gets a ton of downloads and I'm afraid this will break some setups. 
Is it really >=4.9? Not >=4.7?

**Comment by crayfishx** [comment]
Well, it's a soft dependancy - I think the only thing that wouldn't work in earlier 4.x versions is the hiera 5 style data in modules that replaces params.pp which can be circumvented by the user adding some extra Hiera data of their own - not ideal but not impossible either.

It's not 4.7 compatible because it's Hiera 5

What level of version support coverage would you find acceptable?  I'll try and re-work the PR to be compatible with those versions.....

**Comment by crayfishx** [comment]
@roman-mueller I've been trying to get the data in modules to work with Hiera v4 for compatibility but running into some really strange behaviour trying to run the tests on certain versions of Puppet..... v4 Hiera is also experimental and already deprecated (it was the bridge between Hiera 3 and Hiera 5) so I'm not very keen on using it anyway.

I think you need to make a decision on whether you want to accept the new Hiera 5 functionality now or not, and if not then the module will have to continue with the out dated params.pp pattern which would be quite sad:(

Personally, I would release 1.x of the module with the latest supported features (Hiera 5...etc) and comtinue to support 0.x in a branch for the people lagging behind until 4.x goes EOL.   That could be easily documented, and people should expect breaking changes from 0.x to 1.x. It gives the module a chance to not get held back by dated patterns because people don't upgrade but you can still put fixes in 0.x for those people.

Thoughts?

**Comment by ggeldenhuis** [comment]
As a user of this module I would very much support this PR without compromising on removing hiera5 compatibility. Users who use older versions should be using r10k or some other tool to specify a version of the module that they are using. Hiera5 is fact of life and if they want new features they will have to upgrade. It sounds like a hard line but I believe their option to still download an older perfectly functioning version of the module, is a perfectly acceptable compromise.

**Comment by roman-mueller** [comment]
Thanks for the feedback!
I'll merge this and release it was version 1.0.0, I'm going to add a section in the README.md for Puppet <4.9 users.

**Comment by crayfishx** [comment]
Awesome - thanks @roman-mueller !

**Comment by alexjfisher** [comment]
Removing the `operatingsystem` fact from the spec tests turns out not to be a good idea.

Under certain conditions, spec tests will fail with...
```
Evaluation Error: Error while evaluating a Resource Statement, Could not autoload puppet/type/service: Could not autoload puppet/provider/service/bsd: Could not autoload puppet/provider/service/init: undefined method `downcase' for nil:NilClass
```
https://travis-ci.org/voxpupuli/puppet-cron/jobs/311845970#L656

https://github.com/voxpupuli/puppet-cron/pull/33#issuecomment-349466261

File: manifests/init.pp [hunk 1]
```diff
@@ -34,53 +34,98 @@
...
+  $cron_job_multiple.each | String $t, Hash $params | {
+    cron::job::multiple { $t:
+      * => $params
+    }
   }
 
-  $cron_job_hourly = hiera_hash('cron::hourly', undef)
-  if $cron_job_hourly {
-    create_resources('cron::hourly', $cron_job_hourly)
+  $cron_job_hourly = lookup('cron::job::hourly', Optional[Hash], 'hash', {})
```
Comment (alexjfisher) [inline manifests/init.pp]: @crayfishx Is this right?  The previous code did a lookup for `cron::hourly`, the replacement uses `cron::job::hourly`.

File: manifests/init.pp [hunk 2]
```diff
@@ -34,53 +34,98 @@
...
+  $cron_job_hourly.each | String $t, Hash $params | {
+    cron::job::hourly { $t:
```
Comment (alexjfisher) [inline manifests/init.pp]: No such defined type.

**Pull request merged**